pilot/pkg/xds/debuggen.go

```diff
@@ -0,0 +1,142 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/golang/protobuf/ptypes/any"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	// TypeDebug requests debug info from istio, a secured implementation for istio debug interface
+	TypeDebug = "istio.io/debug"
+	debugURL  = "http://localhost"
+)
+
+var activeNamespaceDebuggers = map[string]struct{}{
+	"config_dump": {},
+	"ndsz":        {},
+	"edsz":        {},
+}
+
+// DebugGen is a Generator for istio debug info
+type DebugGen struct {
+	Server          *DiscoveryServer
+	SystemNamespace string
+}
+
+type ResponseCapture struct {
+	body        *bytes.Buffer
+	header      map[string]string
+	wroteHeader bool
+}
+
+func (r ResponseCapture) Header() http.Header {
+	header := make(http.Header)
+	for k, v := range r.header {
+		header.Set(k, v)
+	}
+	return header
+}
+
+func (r ResponseCapture) Write(i []byte) (int, error) {
+	return r.body.Write(i)
+}
+
+func (r ResponseCapture) WriteHeader(statusCode int) {
+	r.header["statusCode"] = strconv.Itoa(statusCode)
+}
+
+func NewResponseCapture() *ResponseCapture {
+	return &ResponseCapture{
+		header:      make(map[string]string),
+		body:        new(bytes.Buffer),
+		wroteHeader: false,
+	}
+}
+
+func NewDebugGen(s *DiscoveryServer, systemNamespace string) *DebugGen {
+	return &DebugGen{
+		Server:          s,
+		SystemNamespace: systemNamespace,
+	}
+}
+
+// Generate XDS debug responses according to the incoming debug request
+func (dg *DebugGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource,
+	updates *model.PushRequest) (model.Resources, error) {
+	res := []*any.Any{}
+	var buffer bytes.Buffer
+	if w.ResourceNames == nil {
+		return res, fmt.Errorf("debug type is required")
+	}
+	if len(w.ResourceNames) != 1 {
+		return res, fmt.Errorf("only one debug request is allowed")
+	}
+	resourceName := w.ResourceNames[0]
+	u, _ := url.Parse(resourceName)
+	debugType := u.Path
+	identity := proxy.VerifiedIdentity
+	if identity.Namespace != dg.SystemNamespace {
+		shouldAllow := false
+		if _, ok := activeNamespaceDebuggers[debugType]; ok {
+			shouldAllow = true
+		}
+		if !shouldAllow {
+			return res, fmt.Errorf("the debug info is not available for current identity: %q", identity)
+		}
+	}
+	debugURL := debugURL + debugServerPort + "/debug/" + resourceName
+	req, _ := http.NewRequest(http.MethodGet, debugURL, nil)
+	handler, _ := debugmux.Handler(req)
+	response := NewResponseCapture()
+	handler.ServeHTTP(response, req)
+	if response.wroteHeader && len(response.header) >= 1 {
```

Comment:
I think we can create a struct like
```go
type DebugResponse struct{
  Metadata map[string][]string
  Body []byte
}
```

Then we can just do json.marshal(debugResponse) instead of hand writing the json

Reply:
fixed